### PR TITLE
Adding the output formatter

### PR DESCRIPTION
**Pull request by patrick-wright**
Implementing ECO-248

Adding an output formatter which shows duration of each command and whether it failed or not

### CODE REVIEW

File: lib/chef-acceptance/application.rb [hunk 1]
```diff
@@ -34,7 +53,7 @@ def run(suite, command)
             puts "Encountered an error running the recipe #{recipe}: #{e.message}\n#{e.backtrace.join("\n")}"
             if force_destroy && recipe != "destroy"
               puts "--force-destroy specified so attempting to run the destroy recipe"
-              run_command(test_suite, "destroy")
+              run_command(test_suite, "force-destroy")
```
Comment (patrick-wright) [inline lib/chef-acceptance/application.rb]: Can't we just pass "destroy" versus making up a new command?

File: lib/chef-acceptance/application.rb [hunk 2]
```diff
@@ -42,15 +61,20 @@ def run(suite, command)
...
-
-      # After we run all the suites we need to output the duration and return an
-      # non-zero exit code if there was an error
     end
 
     def run_command(test_suite, command)
-      runner = ChefRunner.new(test_suite, command)
-      runner.run!
-      # TODO add statistics to output
+      recipe = command == "force-destroy" ? "destroy" : command
```
Comment (patrick-wright) [inline lib/chef-acceptance/application.rb]: See comment above, do we need to check for "force-destroy" versus just "destroy"?

File: lib/chef-acceptance/application.rb [hunk 3]
```diff
@@ -34,7 +53,7 @@ def run(suite, command)
             puts "Encountered an error running the recipe #{recipe}: #{e.message}\n#{e.backtrace.join("\n")}"
             if force_destroy && recipe != "destroy"
               puts "--force-destroy specified so attempting to run the destroy recipe"
```
Comment (patrick-wright) [inline lib/chef-acceptance/application.rb]: using `puts` in this class doesn't feel right.  Shouldn't Cli be the only class printing to console?

File: chef-acceptance.gemspec [hunk 4]
```diff
@@ -12,8 +12,11 @@ Gem::Specification.new do |s|
   s.executables   = %w(chef-acceptance)
```
Comment (patrick-wright) [inline chef-acceptance.gemspec]: remove this line

File: lib/chef-acceptance/application.rb [hunk 5]
```diff
@@ -4,26 +4,62 @@
...
+      # Make sure that exit code reflects the error if we have any
+      exit(1) if error
+    end
+
+    def run_suite(suite, command)
       test_suite = TestSuite.new(suite)
 
       unless test_suite.exist?
-        fail <<-EOS.gsub(/^\s+/, "")
+        raise AcceptanceError, <<-EOS.gsub(/^\s+/, "")
```
Comment (patrick-wright) [inline lib/chef-acceptance/application.rb]: much better

File: lib/chef-acceptance/output_formatter.rb [hunk 6]
```diff
@@ -1,69 +1,89 @@
...
+    # | Suite  | Command   | Duration   | Error |
+    # | suite1 | provision | 00:00:00   | N     |
+    # | suite1 | verify    | 00:00:01   | Y     |
+    # | suite1 | destroy   | 00:00:10   | N     |
+    # | suite2 | provision | 00:01:40   | N     |
+    # | suite2 | verify    | 00:16:40   | N     |
+    # | suite2 | destroy   | 2777:46:40 | N     |
+    def generate_output
+      # Change duration to a more readable format
+      rows.each do |row|
```
Comment (patrick-wright) [inline lib/chef-acceptance/output_formatter.rb]: combine the `rows.each` blocks?

```ruby
      rows.each do |row|
        # Change duration to a more readable format
        row[2] = pretty_duration(row[2])

        # Error comes in as true or false - change to 'Y' or 'N'
        row[3] = row[3] ? "Y" : "N"
      end
```

File: lib/chef-acceptance/output_formatter.rb [hunk 7]
```diff
@@ -1,69 +1,89 @@
...
-    end
+    private
 
-    def add_error(suite, command, error)
-      stats[suite] ||= {}
-      stats[suite][command] ||= {}
-      stats[suite][command][:error] = error
-    end
+    # Takes in duration as a number of seconds and turns it into a
+    # pretty output like `1:23:59`
```
Comment (patrick-wright) [inline lib/chef-acceptance/output_formatter.rb]: Should be identify each segment? 1:23:59 -> 1h:23m:59s

File: lib/chef-acceptance/application.rb [hunk 8]
```diff
@@ -4,26 +4,62 @@
...
+          output_formatter.add_row(suite: "", command: "Total", duration: total_duration, error: error)
+
+          puts ""
+          puts "chef-acceptance run #{error ? "failed" : "succeeded"}."
+          puts output_formatter.generate_output
+        end
+      end
+
+      # Make sure that exit code reflects the error if we have any
+      exit(1) if error
```
Comment (patrick-wright) [inline lib/chef-acceptance/application.rb]: What do you think about raising an AcceptanceError instead and letting the Cli class handle the exception and call exit(1)?

File: lib/chef-acceptance/output_formatter.rb (hunk 7 above)
Comment (sersut) [inline lib/chef-acceptance/output_formatter.rb]: I'm :-1: on this :smile:

File: chef-acceptance.gemspec (hunk 4 above)
Comment (sersut) [inline chef-acceptance.gemspec]: Why don't we have chef-acceptance listed here @patrick-wright ?

File: lib/chef-acceptance/application.rb (hunk 8 above)
Comment (sersut) [inline lib/chef-acceptance/application.rb]: When we raise something here, Thor prints the exception to stderr pushing our table up in the log. 

We can make the CLI handle the exception, but really nothing should be done after this point. So exception handling logic will be just `exit(1)` which is why I wanted to have it in here.

File: lib/chef-acceptance/output_formatter.rb (hunk 6 above)
Comment (sersut) [inline lib/chef-acceptance/output_formatter.rb]: :+1:

File: spec/integration/cli_spec.rb [hunk 9]
```diff
@@ -2,156 +2,115 @@
...
+    context "with test command" do
+      let(:options) { %w{test error-suite} }
+      let(:failure_expected) { true }
+
+      it "fails" do
+        output = run_acceptance
+        expect(output).to match(/provision phase/)
+        expect(output).to match(/recipes\/verify.rb:1: syntax error/)
+        expect(output).to match(/\| error-suite \| provision \| [\d][\d]:[\d][\d]:[\d][\d] \| N     \|/)
+        expect(output).to match(/\| error-suite \| verify    \| [\d][\d]:[\d][\d]:[\d][\d] \| Y     \|/)
```
Comment (sersut) [inline spec/integration/cli_spec.rb]: There probably is a better regex then `[\d][\d]:[\d][\d]:[\d][\d]` :smile:

File: chef-acceptance.gemspec (hunk 4 above)
Comment (patrick-wright) [inline chef-acceptance.gemspec]: It's a a duplicate since it was re-added below.

File: lib/chef-acceptance/application.rb (hunk 8 above)
Comment (patrick-wright) [inline lib/chef-acceptance/application.rb]: I still consider Application to be an API, and I don't like the idea of APIs raising SystemExit.

File: chef-acceptance.gemspec (hunk 4 above)
Comment (sersut) [inline chef-acceptance.gemspec]: Ah got it :smile:

**Comment by tyler-ball** [comment]
:+1:

**Pull request merged**